### ISSUE TITLE
Problem with tmp directory clean-up introduced in v1.9.7.

### ISSUE THREAD

**Issue by tvaughan** [bundler/bundler#3641]
This happens with v1.9.7 but not v1.9.6. Where "this" is a call to `bundle install` without using sudo. Using sudo does solve this problem too. However, I would like to not have to require sudo. This is on an Ubuntu trusty64 machine in EC2. This happens when we try to provision with ansible over ssh, or ssh into the box and run `bundle install` manually.

The Gemfile in question is:

```ruby
# -*- mode: ruby -*-
source 'https://rubygems.org'

gem 'rake'
gem 'serverspec'

gem 'faraday'
gem 'faraday_middleware'
```

The error looks like:

```
TASK: [serverspec | Use bundler to install rake and serverspec] *************** 
failed: [127.0.0.1] => {"changed": true, "cmd": "bundle install", "delta": "0:00:03.498658", "end": "2015-05-12 18:40:54.992000", "rc": 5, "start": "2015-05-12 18:40:51.493342", "warnings": []}
stderr: mv: cannot stat ‘/tmp/bundler20150512-3547-rqcczrrake-10.4.2/cache/rake-10.4.2.gem’: No such file or directory
stdout: Fetching gem metadata from https://rubygems.org/........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..

Gem::Exception: Cannot load gem at [/var/lib/gems/1.9.1/cache/rake-10.4.2.gem] in /srv/search/serverspec
An error occurred while installing rake (10.4.2), and Bundler cannot continue.
Make sure that `gem install rake -v '10.4.2'` succeeds before bundling.

FATAL: all hosts have already failed -- aborting
```

**Comment by indirect** [comment]
Yikes, that's not good. Thanks for reporting this! Please downgrade to 1.9.6 for now, and we'll try to have a fix for this problem released within the next day or two.

**Comment by tvaughan** [comment]
> Please downgrade to 1.9.6 for now

Yup. Done.

> we'll try to have a fix for this problem released within the next day or two.

Thank you sir.

**Comment by hackeron** [comment]
I'm wondering if this is related to the problem I am having, which is only present in 1.9.7, but not in 1.9.6: https://gist.github.com/hackeron/687adce4862c1a45991b

**Comment by jfly** [comment]
I'm seeing the same thing over here. Dumping my investigation from the last hour in case anyone finds it useful.

The problem seems to be due to this call to [mv](https://github.com/bundler/bundler/blob/1-9-stable/lib/bundler/fetcher.rb#L67):

```ruby
Bundler.sudo "mv #{Bundler.tmp(spec.full_name)}/cache/#{spec.full_name}.gem #{gem_path}"
```

The immediate problem is that [Bundle.tmp](https://github.com/bundler/bundler/blob/1-9-stable/lib/bundler.rb#L215) generates a new, random, *empty* folder each time, so this `mv` is always going to fail. It seems to me that the fix would be to change the problematic mv to

```ruby
Bundler.sudo "mv #{download_path}/cache/#{spec.full_name}.gem #{gem_path}"
```

A more general concern is that this call to Bundler.sudo failed, but we continued to limp along. I personally feel it would be better for Bundler.sudo to fail-fast. Digging into the call to Bundler.sudo, I see that it [calls](https://github.com/bundler/bundler/blob/1fe23219d42365693326b716eb3d9a50973a4137/spec/support/sudo.rb#L9) [sys_exec](https://github.com/bundler/bundler/blob/1fe23219d42365693326b716eb3d9a50973a4137/spec/support/helpers.rb#L135). sys_exec merely prints out an error if there was one, instead of failing fast:

```ruby
      puts @err unless expect_err || @err.empty? || !$show_err
      @out
    end
```

**Comment by indirect** [comment]
Yeah, I think that's correct. How did our sudo specs not catch this? 😧

As for `sys_exec`, printing to `stderr` definitely does not mean the command failed. If we are going to fail fast, we need to use the exit status of the command, which is only available on (iirc) Ruby 2.0 and up.

**Comment by jfly** [comment]
Ah, good point. I wasn't reading the code very carefully. Definitely should be looking at the exit code if we want to fail fast.

**Comment by indirect** [comment]
Fixed in 1.9.8, just released now!

**Comment by segiddins** [comment]
Closed by https://github.com/bundler/bundler/issues/3642.
